### ISSUE TITLE
Reveal body background issue (Firefox / 6.4.4-rc1)

### ISSUE THREAD

**Issue by lukeyouell** [zurb/foundation-sites#10791]
#### How to reproduce this bug:

1. Create a reveal as per docs

#### What should happen:

1. Reveal shouldn't affect other elements of the page

#### What happened instead:

1. Upon opening the reveal modal, the <body> background is changed to white

#### Browser(s) and Device(s) tested on:

- Latest version of Firefox

#### Foundation Version(s) you are using:

6.4.4-rc1 (the issue doesn't exist on 6.4.3)

#### Test case link:

https://codepen.io/lukeyouell/pen/VrGvQP

**Comment by ncoden** [comment]
This does not come from Reveal, but from a browser bug in Firefox:
* with `html { position: fixed; }`, the html is collapsed and no content has to be print
* with `html { width: 100%; }`, the html take the full width, but the body does not print its gradient
* with `body { background: #000; }`, the background is correctly displayed

See: https://codepen.io/ncoden/pen/dJvoqq

**Comment by ncoden** [comment]
I reported this issue at https://bugzilla.mozilla.org/show_bug.cgi?id=1427251

**Comment by ncoden** [comment]
I think this was introduced at https://github.com/zurb/foundation-sites/pull/10583, where the `position: fixed` trick was applied to all browser instead of only iOS. The thing is, we had a lot of troube to prevent scroll in modal in all browsers, and I see no way to fix the current issue without high probability of breaking it.

So for now (waiting for https://github.com/zurb/foundation-sites/issues/10824), I'll go for the simplest option which is applying the gradient background on `<html>` or any child `<div>` instead of on `<body>`.

What do you think @lukeyouell ?

Poke @isapir

**Comment by isapir** [comment]
I'm sorry, when I looked yesterday I didn't realize that there are 2 html elements in CodePen, one from CodePen website and one from the pen itself.  I will have to look at it with a clear mind.  My mind is currently deep in other code.

What I don't understand though, is how come we didn't see it in the Reveal docs when I submitted the patch?  @kball also confirmed that my patch worked.

**Comment by ncoden** [comment]
@isapir

**Comment by ncoden** [comment]
Hi @isapir. Will you have some time to look at this ?

**Comment by isapir** [comment]
@ncoden @DanielRuf The first thing that I see is that #11065 broke my fix for #7831 on Chrome.  While there might be a browser bug in Firefox that causes another issue, I don't think that this fix was right.  Perhaps we should check if a scrollbar is needed or not by comparing the client height with the content's height.

As you can see in the screenshots of Chromium, while an additional vertical "disabled scrollbar" is added on the right, the background does not shift, and the problem is therefore resolved:

Screenshots from commit https://github.com/zurb/foundation-sites/commit/7f9e3eda:

![screenshot from 2018-04-01 22-33-15](https://user-images.githubusercontent.com/885907/38184510-009b3724-35fd-11e8-9bb4-68cb17e23b10.png)

![screenshot from 2018-04-01 22-33-22](https://user-images.githubusercontent.com/885907/38184514-09442016-35fd-11e8-92ab-37d5889fde07.png)

But after commit https://github.com/zurb/foundation-sites/commit/107c9ab3 the problem is back on Chrome, and the content below the modal shifts when the Reveal opens:

![screenshot from 2018-04-01 22-44-43](https://user-images.githubusercontent.com/885907/38184698-520e3024-35fe-11e8-9f35-362a7fc7d7d1.png)

![screenshot from 2018-04-01 22-44-47](https://user-images.githubusercontent.com/885907/38184712-62056498-35fe-11e8-8dd3-d16f7bbe02b9.png)

I think that commit 107c9ab3 should be reverted, and then we can find a solution for Firefox until the browser bug is fixed.

**Comment by isapir** [comment]
@ncoden I checked my patch against Firefox and it seems to work properly there as well for #7831 

Firefox screenshots from commit https://github.com/zurb/foundation-sites/commit/7f9e3eda :

![screenshot from 2018-04-01 23-07-31](https://user-images.githubusercontent.com/885907/38185197-e34d8cd0-3601-11e8-9360-00f03b1fec12.png)

![screenshot from 2018-04-01 23-07-37](https://user-images.githubusercontent.com/885907/38185204-ebfaa32c-3601-11e8-9d63-f456a4349cdb.png)

And in Firefox https://github.com/zurb/foundation-sites/commit/107c9ab broke it as well - notice the shift both on the vertical and horizontal axes:

![screenshot from 2018-04-01 23-12-52](https://user-images.githubusercontent.com/885907/38185277-4d7de56e-3602-11e8-8c3b-cbe337c68e0f.png)

![screenshot from 2018-04-01 23-12-56](https://user-images.githubusercontent.com/885907/38185292-568658f8-3602-11e8-83a8-37513cf8deb9.png)

**Comment by DanielRuf** [comment]
When the reveal is open it should block scrolling on the elements behind it and remove the scrolling from them. The "shift" is because the scrollbars are removed then. A double vertical scrollbar is wrong.

**Comment by DanielRuf** [comment]
> find a solution for Firefox until the browser bug is fixed

So there is a Firefox bug? Can you file a bugreport at bugzilla of Mozilla so we can track this?

**Comment by ncoden** [comment]
@DanielRuf See  https://bugzilla.mozilla.org/show_bug.cgi?id=1427251

**Comment by DanielRuf** [comment]
I guess two different things were confused, the scrollbar issue and the background issue. @isapir

**Comment by isapir** [comment]
> When the reveal is open it should block scrolling on the elements behind it and remove the scrolling from them. 

I just tested again and you can not scroll when the modal is open.  The scrollbar is disabled.  @DanielRuf @ncoden

**Comment by ncoden** [comment]
@isapir Your patch works, this is not the point. The point is that there is a double scrollbar in Firefox due to the two `overflow: scroll`. Does your patch need two of them ?

**Comment by DanielRuf** [comment]
`scroll` should be at least `auto` or many browsers always show a acrollbar even if there is nothing to scroll.

**Comment by isapir** [comment]
> The point is that there is a double scrollbar in Firefox due to the two overflow: scroll. Does your patch need two of them ?

@ncoden I really don't remember all of the details, but at the time I researched the issue in depth and that was the only viable solution that I found.  As you can see in the comments of #7831 that issue was opened for a long time with no good solution, and it affected many users.

I don't think that the patch adds 2 scrollbars.  If you look for example at https://foundation.zurb.com/sites/docs/reveal.html (screenshot below) then you can see that there are already two scrollbars on the right side.

> scroll should be at least auto or many browsers always show a acrollbar even if there is nothing to scroll.

@DanielRuf We can perhaps check the height of the content in JavaScript and show/hide the scrollbar if the content is shorter than the window.

![image](https://user-images.githubusercontent.com/885907/38205295-f70746d6-365a-11e8-95d0-4ce2e1281dac.png)

**Comment by DanielRuf** [comment]
> We can perhaps check the height of the content in JavaScript and show/hide the scrollbar if the content is shorter than the window.

This is what `overflow: auto` does, no need for a JavaScript check here.

**Comment by isapir** [comment]
> This is what overflow: auto does, no need for a JavaScript check here.

You convinced me, but can you please tell that to the browsers? ;)

The fact is that the patch that you made broke the fix, so now #7831 does not work anymore.

**Comment by ncoden** [comment]
We could at least try some more test with the isapir patch back, but with one of the two `overflow` set to `auto` ? What do you say @isapir @DanielRuf

**Comment by isapir** [comment]
@ncoden Yes, that would be a good start in my opinion.  At least we'll have a working base point.

**Comment by DanielRuf** [comment]
https://codepen.io/DanielRuf/pen/XWWNRyw

I think we have to wait longer for the upstream fix in Firefox.
